### ISSUE TITLE
px - string or categorical integer values considered as numeric in strip/box/violin plots

### ISSUE THREAD

**Issue by DrGFreeman** [plotly/plotly.py#1918]
#### Issue:

When using integer values as categorical variable in a strip / box / violin plot, the values of the categorical variable are mapped to a continuous numeric axis even if the values are of string or `pd.Categorical` type.

#### Example:

We create a dataframe with columns names having an integer value in string format. These could be any category that makes sense to the specific business case (e.g. product code, etc.)

```python
import numpy as np
import pandas as pd
import plotly_express as px

n = 50

df = pd.DataFrame({
    '1': np.random.normal(2, .3, n),
    '2': np.random.lognormal(.5, .2, n),
    '34': np.random.triangular(0, 2, 3, n),
    '123': np.random.uniform(1, 3, n)
})
```
We unpivot the data using and make a strip plot.

```python
df1 = df.melt()

px.strip(df1, x='variable', y='value')
```
![string](https://user-images.githubusercontent.com/22901938/66615416-9c394280-eb9a-11e9-87e6-b0326c1f8053.png)

The categorical variable (values '1', '2', '34' and '123') get mapped to a continuous numeric scale. Here, the variables '1' and '2' blend together and this can get worst if there are orders of magnitude between the different values.

Converting the string values to `pd.Categorical` type yields the same result as above.

```python
df2 = df.melt()
df2.variable = pd.Categorical(df2.variable)

px.strip(df2, x='variable', y='value')
```
#### Workaround:

Adding a character to the values makes them be recognized as categorical which is the expected result (except for the added character in the category names). Unfortunately, adding a blank space does not work either.
```python
df3 = df.copy()
df3.columns = [f"c{c}" for c in df3.columns]

px.strip(df3.melt(), x='variable', y='value')
```
![string+char](https://user-images.githubusercontent.com/22901938/66615486-ec180980-eb9a-11e9-8a6a-521a38bd2841.png)

Considering that numeric categorical values are legitimate in many contexts, it should be possible to use numbers as categories if they are represented by a string or categorical data type, as is the case with the `color` parameter (https://github.com/plotly/plotly_express/issues/140):

```python
px.strip(df.melt(), y='value', color='variable')
```
![color](https://user-images.githubusercontent.com/22901938/66616103-795c5d80-eb9d-11e9-9970-626ff82f68a0.png)

Thanks!

```
Package              Version  
-------------------- --------- 
plotly               4.1.1    
plotly-express       0.4.1    
```

@emmanuelle, this is one of the two issues we discussed at the PyData meetup.

**Comment by emmanuelle** [comment]
Hey @DrGFreeman sure this is a valid concern. You can force the axis to be categorical by creating the plotly figure using the `px` function and then do 
```
fig.update_layout(xaxis_type='category')
```
which will force your axis to be categorical. Then of course there is the question whether we should impose this at the `plotly.express` level...

**Comment by DrGFreeman** [comment]
Thanks for the quick response @emmanuelle. I will use this tip for sure.

> Then of course there is the question whether we should impose this at the plotly.express level...

Of course, that is up to the core developers to decide. From the perspective of API consistency, if integers passed as strings to the `color` parameter yield a categorical color scale (as opposed to a continuous one), I would intuitively expect to get similar results on an axis.

**Comment by cvrnogueira** [comment]
@emmanuelle  thanks for your solution, but when we are in a plot using `facet_col` this just work for the last plot, not for all of them. Do you have any solution? Thanks

**Comment by nicolaskruchten** [comment]
@cvrnogueira If you use `fig.update_xaxes(type='category')` it will apply to all facets.

**Comment by nicolaskruchten** [comment]
> Of course, that is up to the core developers to decide. From the perspective of API consistency, if integers passed as strings to the color parameter yield a categorical color scale (as opposed to a continuous one), I would intuitively expect to get similar results on an axis.

I agree, but for backwards-compatibility reasons throughout the API, we can't really change this at the moment: we've always accepted stringified numbers as numbers on positional axes like `x` and `y`

**Comment by harrybiddle** [comment]
This is also an issue with `Figure.add_shape`. Even when the axis has been set to be categorical and the x0/x1 is of string type, if it represents a number it is still interpreted as a coordinate rather than a category. I've searched through the API and I can't find a way around this... Any suggestions welcome.
